pkg/sif/write.go

```diff
@@ -145,9 +145,73 @@ func (f *fileImage) writeIndexToFileImage(ii v1.ImageIndex, rootIndex bool) erro
 	return f.writeBlobToFileImage(bytes.NewReader(m), rootIndex)
 }
 
+// numDescriptorsForImage returns the number of descriptors required to store img.
+func numDescriptorsForImage(img v1.Image) (int64, error) {
+	ls, err := img.Layers()
+	if err != nil {
+		return 0, err
+	}
+
+	return int64(len(ls) + 2), nil
+}
+
+// numDescriptorsForIndex returns the number of descriptors required to store ii.
+func numDescriptorsForIndex(ii v1.ImageIndex) (int64, error) {
+	index, err := ii.IndexManifest()
+	if err != nil {
+		return 0, err
+	}
+
+	var count int64
+
+	for _, desc := range index.Manifests {
+		//nolint:exhaustive // Exhaustive cases not appropriate.
+		switch desc.MediaType {
+		case types.DockerManifestList, types.OCIImageIndex:
+			ii, err := ii.ImageIndex(desc.Digest)
+			if err != nil {
+				return 0, err
+			}
+
+			n, err := numDescriptorsForIndex(ii)
+			if err != nil {
+				return 0, err
+			}
+
+			count += n
+
+		case types.DockerManifestSchema2, types.OCIManifestSchema1:
+			img, err := ii.Image(desc.Digest)
+			if err != nil {
+				return 0, err
+			}
+
+			n, err := numDescriptorsForImage(img)
+			if err != nil {
+				return 0, err
+			}
+
+			count += n
+
+		default:
+			count++
+		}
+	}
+
+	return count + 1, nil
+}
+
 // Write constructs a SIF at path from an ImageIndex.
 func Write(path string, ii v1.ImageIndex) error {
-	fi, err := sif.CreateContainerAtPath(path, sif.OptCreateDeterministic())
+	n, err := numDescriptorsForIndex(ii)
+	if err != nil {
+		return err
+	}
+
+	fi, err := sif.CreateContainerAtPath(path,
+		sif.OptCreateDeterministic(),
+		sif.OptCreateWithDescriptorCapacity(n),
```

Comment:
What's the route if we need to have additional spare descriptors? E.g. to allow adding an embedded overlay as in https://github.com/sylabs/singularity/issues/1480

Reply:
There isn't one currently. The naive way of doing that would be to update the `ImageIndex`, and call `Write` with your updated `ImageIndex`, passing a new path. Obviously that has some downsides in terms of performance and resource usage.

We could add an option to `Write` that would specify a number of extra descriptors, but it might be difficult for Singularity to determine what a suitable value would be? At a minimum, adding an overlay would require a new descriptor for the `extfs` image itself, and if we want to update the image manifest/config and root index to make things consistent, that would require _three more_ descriptors _per modification_.

Taking a step back though, another approach to support https://github.com/sylabs/singularity/issues/1480 would be a way of providing an updated `ImageIndex`, and having it figure out how to reconcile the SIF by:

- Garbage collect blob(s) that are no longer referenced.
- Extend the number of descriptors in the SIF (if necessary.)
- Add blob(s) that are not present.

Thoughts?

Reply:
I think we need to talk through this a bit. It's fairly important for HPC use cases that we aren't having to Write huge images out to temporary paths etc. frequently. Any time we increase the usage of temp space, or read->write things we tend to get complaints.

I guess I was assuming that if the ImageIndex and RootIndex could be at the end of the OCI-SIF, and we had spare descriptors in the header, then we could _rewrite_ the ImageIndex, RootIndex and add the extfs while leaving all the existing layer blobs intact... no need to add new descriptors for the updated ImageIndex and RootIndex.

A separate operation to later make all the hashes consistent after the extfs has been modified may require rewriting the OCI-SIF in full.

Reply:
> I think we need to talk through this a bit. It's fairly important for HPC use cases that we aren't having to write huge images out to temporary paths etc. frequently. Any time we increase the usage of temp space, or read->write things we tend to get complaints.

Yeah, no argument there! The point I was trying to make was that `Write` might not be expressive enough to capture the modify-in-place intent, and figure out what its role in supporting that might be.

I could see `Write` having an "extra descriptors" option to make the modify-in-place flow more efficient... and I think this was what you were alluding to?

> I guess I was assuming that if the ImageIndex and RootIndex could be at the end of the OCI-SIF, and we had spare descriptors in the header, then we could rewrite the ImageIndex, RootIndex and add the extfs while leaving all the existing layer blobs intact... no need to add new descriptors for the updated ImageIndex and RootIndex. (given we'd be deleting and addding things only from the end of the SIF...?)
>
> A separate operation to later make all the hashes consistent after the extfs has been modified may require rewriting the OCI-SIF in full.

Yeah, in general the filesystem blob(s) are written before the image manifest/config and root index, so we could get away without re-writing the filesystem blob(s). And I would suggest you probably want overlay blob(s) to end up before the image manifest/config and root index to make the hash consistency operation to be more efficient?

We probably can't rewrite the image manifest/config and root index in place (they will grow when more blobs are added), but we could:

- Delete the `ImageIndex`, `ImageConfig` and `RootIndex` blobs (compacting the image.)
- Add the overlay blob.
- Re-add the `ImageIndex`, `ImageConfig` and `RootIndex` blobs.

Then to make the hashes consistent, all you'd need to do would be:

- Delete the `ImageIndex`, `ImageConfig` and `RootIndex` blobs (compacting the image.)
- Re-add the `ImageIndex`, `ImageConfig` and `RootIndex` blobs.

Probably worth speccing this out in issues here? From what I can see:

- Add `Write` option that allows a user to specify a number of extra descriptors to include in the image.
- Spec out one or more API(s) to support the addition of an overlay, and to update the relevant image manifest/config and root index after an overlay is modified.

Reply:
:+1: Yep, that's the layout and process I was assuming... always keep the Index / Config stuff at the end, after layer and overlay blobs.